srcStatic/ModuleLoader.cpp

```diff
@@ -1,58 +1,27 @@
 #include "ModuleLoader.h"
-#include "WindowsErrorCode.h"
-#include "FileSystemHelper.h"
+#include "GameModules/IniModule.h"
 #include "StringConversion.h"
-#include "GlobalDefines.h"
-#include <stdexcept>
+#include "FileSystemHelper.h"
 
-// Load all active modules specified in the .ini file
-void ModuleLoader::LoadModules()
-{
-	const auto moduleNames = GetModuleNames("ExternalModules");
 
-	for (const auto& moduleName : moduleNames)
-	{
-		LoadModule(moduleName);
-	}
-}
-
-void ModuleLoader::LoadModule(std::string sectionName)
+ModuleLoader::ModuleLoader() 
 {
-	IniModuleEntry moduleEntry;
-	moduleEntry.iniSectionName = sectionName;
-
-	try {
-		moduleEntry.handle = LoadModuleDll(sectionName);
-	}
-	catch (const std::exception& error) {
-		PostErrorMessage(__FILE__, __LINE__, error.what());
-		return;
-	}
-
-	CallModuleInitialization(moduleEntry, sectionName);
-
-	moduleEntry.destroyModFunc = (DestroyModFunc)GetProcAddress(moduleEntry.handle, "DestroyMod");
-
-	modules.push_back(moduleEntry);
+	RegisterExternalModules();
 }
 
-// Unload all active modules specified in the .ini file
-bool ModuleLoader::UnloadModules()
+void ModuleLoader::RegisterExternalModules()
 {
-	bool result = true;
+	const auto sectionNames = GetModuleNames("ExternalModules");
 
-	for (IniModuleEntry moduleEntry : modules)
+	for (const auto& sectionName : sectionNames)
 	{
-		if (!CallModuleDestruction(moduleEntry)) {
-			result = false;
+		try {
+			RegisterModule(static_cast<std::unique_ptr<GameModule>>(std::make_unique<IniModule>(sectionName)));
```

Comment:
Hmm, right, `unique_ptr` doesn't play very well with inheritance. I'm kind of thinking we should go with convention about ownership, rather than enforcing strict `unique_ptr` semantics. It would also allow us to keep using reference syntax instead of pointer syntax.

Reply:
As in, we assume the module loader has ownership of all registered modules, and it's free to delete / destruct those objects. The compiler won't try to clear out the pointers of a caller when transferring ownership, though if the caller happened to hang on to a pointer, it may risk a crash if it called methods after the `ModuleLoader` had destructed that module. That's not to say a stored module pointer could never be used, just we'd need to have a clear definition of when the `ModuleLoader` would destroy the object to determine when it may be safe to call methods on the module.

```
auto module = new SomeGameModule();
module.Method(); // Sure, this is fine.

moduleLoader.Register(module); // Transfer ownership to module loader (by convention)

module.Method(); // This might be unsafe, if moduleLoader has destroyed the object

moduleLoader.UnloadModules(); // Though the module loader might specify it's safe until now

module.Method(); // Undefined behavior. Likely a crash. Method called on destroyed object

delete module; // Bad! You don't have ownership. This is moduleLoader's responsibility.
```

More likely though, the code could end up looking something like:
```
moduleLoader.RegisterModule(new SomeGameModule());
// No local pointer available to call methods, or potentially do anything bad with
```

Reply:
If we want to allow access to the module outside of ModuleLoader after it has been registered, we could use a shared_ptr. I wasn't sure we wanted to go with a unique_ptr, but it seemed to be the easier to test and maintain solution so I went with it. If we want modules to be manipulated after registration, I wouldn't be opposed, but was sort of waiting to see if a use case appeared before supporting it.

I would prefer a shared or unique pointer to a raw pointer because they tell the user more about the intention and working of the pointer without having to review comments or conventions. It also reduces the chance of trying to de-allocate memory incorrectly by allowing the standard library to cover it.

I suppose if it was proven that we could not accomplish our goals for how the ModuleLoader works using a shared or unique pointer, I would be amenable to using a raw pointer, but I haven't seen that yet.

Reply:
Currently the use of `unique_ptr` requires some pretty ugly casts at the point of call. That's what I really want to avoid. I don't think we really need to switch to `shared_ptr`, as that would leave the problem the same.

I'm not opposed to using `unique_ptr` internally to manage lifetime. I'm mostly just concerned about the impact it has on the API. I would like an API where the caller doesn't have to cast to call a method.

Supporting manipulation of the modules after loading doesn't seem particularly important. It's more that not using a `unique_ptr` makes that a possibility that should be considered.

----

Bit of a side note here.

There is also another API cost to using `unique_ptr`, which ties in with taking ownership of objects (whether or not they're using `unique_ptr`). When you take ownership in a way that guarantees `delete` will be called on the object pointer at some point, you can no longer pass global or local variables. It would be an error to try to `delete` a global or a local variable.

The way some of the code is currently written, a few of our patches are written as global objects. That is however the ad hoc nature of existing patch code. As we rewrite things, that will change. Hence, this is just a side note, not really a show stopper.

With the current plan going forward, particularly if we use RAII and apply patches in the constructors and remove them in destructors, it does make sense to create the patches as dynamic objects and fully manage their lifetime.

Reply:
I found this article, which may be relevant here:
[Move smart pointers in and out functions in modern C++](https://www.internalpointers.com/post/move-smart-pointers-and-out-functions-modern-c)

It includes some guidelines for when to use smart pointers in interfaces, including a link to a [C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#r30-take-smart-pointers-as-parameters-only-to-explicitly-express-lifetime-semantics) document, which examples of when and why not to.

I've been somewhat aware there were downsides, and have been hesitant to use them as part of my own interfaces, though I never really read in detail why, nor assessed if this here is an appropriate use case.

Reply:
This does actually seem like an appropriate use of `unique_ptr` as a function parameter.

The real issue with the casting came from pass by reference, rather than pass by value.